crud/common/call.lua

```diff
@@ -12,14 +12,74 @@ local call = {}
 
 local DEFAULT_VSHARD_CALL_TIMEOUT = 2
 
-local function call_impl(vshard_call, func_name, func_args, opts)
-    dev_checks('string', 'string', '?table', {
+
+-- for testing
+call.internal = {}
+
+local function get_vshard_call_name(mode, prefer_replica, balance)
+    dev_checks('string', '?boolean', '?boolean')
+
+    if mode == 'write' then
+        return 'callrw'
+    end
+
+    if not prefer_replica and not balance then
+        return 'callro'
+    end
+
+    if not prefer_replica and balance then
+        return 'callbro'
+    end
+
+    if prefer_replica and not balance then
+        return 'callre'
+    end
+
+    -- prefer_replica and balance
+    return 'callbre'
+end
+
+local function wrap_vshard_err(err, func_name, replicaset_uuid, bucket_id)
+    if err.type == 'ClientError' and type(err.message) == 'string' then
+        if err.message == string.format("Procedure '%s' is not defined", func_name) then
+            if func_name:startswith('_crud.') then
+                err = NotInitializedError:new("crud isn't initialized on replicaset: %q", replicaset_uuid)
+            else
+                err = NotInitializedError:new("Function %s is not registered", func_name)
+            end
+        end
+    end
+
+    if replicaset_uuid == nil then
+        local replicaset, _ = vshard.router.route(bucket_id)
+        if replicaset == nil then
+            return CallError:new(
+                "Function returned an error, but we couldn't figure out the replicaset: %s", err
+            )
+        end
+
+        replicaset_uuid = replicaset.uuid
+    end
+
+    err = errors.wrap(err)
+
+    return CallError:new(utils.format_replicaset_error(
+        replicaset_uuid, "Function returned an error: %s", err
+    ))
+end
+
+function call.map(func_name, func_args, opts)
+    dev_checks('string', '?table', {
+        mode = 'string',
+        prefer_replica = '?boolean',
```

Comment:
Maybe `mode='write' + prefer_replica=true` should cause an error.

But feel free to ignore.

Reply:
`vshard` ignores `prefer_replica` and `balance` options if `write` mode is specified. Let's do the same.